File: leoz-central/src/main/java/org/deku/leoz/central/messaging/IdentityMessageHandler.java
package org.deku.leoz.central.messaging;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leoz.central.data.entities.jooq.tables.records.MstNodeRecord;
import org.deku.leoz.central.data.repositories.NodeRepository;
import org.deku.leoz.node.messaging.auth.v1.AuthorizationMessage;
import org.deku.leoz.node.messaging.auth.v1.IdentityMessage;
import sx.jms.Handler;
import sx.jms.Converter;
import sx.jms.converters.DefaultConverter;

import javax.jms.*;

/**
 * Created by masc on 01.07.15.
 */
public class IdentityMessageHandler implements Handler<IdentityMessage> {
    private Log mLog = LogFactory.getLog(this.getClass());

    private NodeRepository mNodeRepository;
    private Converter mConverter;

    public IdentityMessageHandler(NodeRepository nodeRepository) {
        mNodeRepository = nodeRepository;

        mConverter = new DefaultConverter(
                DefaultConverter.SerializationType.KRYO,
                DefaultConverter.CompressionType.GZIP);
    }

    @Override
    public void onMessage(IdentityMessage message, Message jmsMessage, Session session) throws JMSException {
        try {
            mLog.info(message);

            MstNodeRecord r = mNodeRepository.findByKeyOrCreateNew(message.getKey());

            r.setHostname(message.getHardwareAddress());
            r.setKey(message.getKey());
            r.setSysInfo(message.getSystemInfo());
            r.store();

            Destination replyTo = jmsMessage.getJMSReplyTo();
            if (replyTo != null) {
                AuthorizationMessage am = new AuthorizationMessage();
                am.setId(r.getNodeId());
                am.setKey(r.getKey());
                am.setAuthorized(r.getAuthorized() != null && r.getAuthorized() != 0);

                MessageProducer mp = session.createProducer(replyTo);
                mp.setDeliveryMode(DeliveryMode.NON_PERSISTENT);
                mp.setTimeToLive(10 * 1000);
                mp.setPriority(8);
                mp.send(mConverter.toMessage(am, session));

                session.commit();

                mLog.info(String.format("Sent authorization [%s]", am));
            }
        } catch(Exception e) {
            mLog.error(e.getMessage(), e);
        }
    }
}

File: leoz-central/src/main/java/org/deku/leoz/central/messaging/handlers/IdentityMessageHandler.kt
package org.deku.leoz.central.messaging.handlers

import org.apache.commons.logging.LogFactory
import org.deku.leoz.central.data.repositories.NodeRepository
import org.deku.leoz.node.messaging.entities.AuthorizationMessage
import org.deku.leoz.node.messaging.entities.IdentityMessage
import sx.jms.Channel
import sx.jms.Converter
import sx.jms.Handler
import sx.jms.converters.DefaultConverter
import javax.jms.ConnectionFactory
import javax.jms.Message
import javax.jms.Session

/**
 * Created by masc on 01.07.15.
 */
class IdentityMessageHandler(private val mNodeRepository: NodeRepository) : Handler<IdentityMessage> {
    private val log = LogFactory.getLog(this.javaClass)
    private val converter: Converter

    init {
        converter = DefaultConverter(
                DefaultConverter.SerializationType.KRYO,
                DefaultConverter.CompressionType.GZIP)
    }

    override fun onMessage(message: IdentityMessage, converter: Converter, jmsMessage: Message, session: Session, connectionFactory: ConnectionFactory) {
        try {
            log.info(message)

            var r = mNodeRepository.findByKeyOrCreateNew(message.key)

            r.hostname = message.hardwareAddress
            r.key = message.key
            r.sysInfo = message.systemInfo
            r.store()
            r = mNodeRepository.findByKeyOrCreateNew(message.key)

            val replyTo = jmsMessage.jmsReplyTo
            if (replyTo != null) {
                // Create authorization message
                val am = AuthorizationMessage()
                am.id = r.nodeId
                am.key = r.key
                am.authorized = r.authorized != null && r.authorized !== 0

                Channel(connectionFactory = connectionFactory,
                        jmsSessionTransacted = false,
                        destination = replyTo,
                        converter = this.converter).use { c ->
                    c.send(am)
                }

                log.info("Sent authorization [%s]".format(am))
            }
        } catch (e: Exception) {
            log.error(e.message, e)
        }
    }
}
